### PR TITLE
Add script and update `build-and-test-dbt` workflow to push dbt Python dependencies to S3

### PR DESCRIPTION
**Pull request by jeancochrane**
This PR adds a new script `deploy_dbt_model_dependencies.sh` that downloads dependencies for dbt Python models, bundles them into zip archives, and pushes them to S3. It also updates the `build-and-test-dbt` workflow to use the new script, including the following features:

* If a dbt cache exists, only deploy packages for models that are new or have changed
* If a caller has dispatched the workflow for a specific set of models, only deploy packages for those models
* In all cases, only deploy package bundles that are new or have changed
* Deploy a separate package bundle for each model
* Deploy package bundles to separate S3 locations for dev, CI, and prod, and isolate dev/CI environments by user or branch

Note that this PR leaves out of scope the additional step to refactor the existing `reporting.ratio_stats` model to use this new dependency management system. That step will be handled in a follow-up issue (https://github.com/ccao-data/data-architecture/issues/439).

Closes https://github.com/ccao-data/data-architecture/issues/417.

## Testing

Evidence of `build-and-test-dbt` pushing packages for changed models to S3 on `synchronize` event: [Link](https://github.com/ccao-data/data-architecture/actions/runs/9021296669/job/24788308838#step:7:102)

Evidence of `build-and-test-dbt` skipping dependencies that have not changed on `synchronize` event: [Link](https://github.com/ccao-data/data-architecture/actions/runs/9021342661/job/24788460993#step:7:53)

Evidence of `build-and-test-dbt` skipping models that have no Python dependencies on `workflow_dispatch` event: [Link](https://github.com/ccao-data/data-architecture/actions/runs/9022831999/job/24793287938#step:7:55)

Evidence of `cleanup-dbt-resources` removing packages from S3 on `closed` event: [Link](https://github.com/ccao-data/data-architecture/actions/runs/9021452274/job/24788819738#step:5:35)

### CODE REVIEW

**Comment by jeancochrane** [comment]
Closing this to test the cleanup step.

File: dbt/macros/get_s3_dependency_dir.sql [hunk 1]
```diff
@@ -0,0 +1,48 @@
+-- Get the S3 location that is used to store Python model dependencies
+-- for the current target
```
Comment (jeancochrane) [inline dbt/macros/get_s3_dependency_dir.sql:2]: Ideally we would just add a new attribute `s3_dependency_dir` to our profiles defined in `profiles.yml`, and then we could have full control over the location where dependencies get stored for each target. That doesn't appear to be possible without extending the `dbt-athena` plugin itself to support a new profile variable, so this macro represents a workaround that allows us to compute a dependency dir based on the `s3_data_dir` variables set in our profiles. 

I think we'll probably need a Python version of this macro as well, but I'm leaving that for myself to figure out later when I tackle https://github.com/ccao-data/data-architecture/issues/439.

File: dbt/macros/get_s3_dependency_dir.sql [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+                return(
+                    raise_error_func(
+                        "HEAD_REF env var must be set when target is 'ci'"
+                    )
+                )
+            }}
+        {% endif %}
+        {% set dir_suffix = "/" ~ head_ref %}
+    {% endif %}
+    {% set s3_dependency_dir = target.s3_data_dir ~ "packages" ~ dir_suffix %}
```
Comment (jeancochrane) [inline dbt/macros/get_s3_dependency_dir.sql:40]: One downside of this approach is that it stores packages _inside_ the data dir, when ideally they would be stored alongside the data dir. For clarity, here are the three paths for the three different targets we support:

* `dev`
  * data dir:  `s3://ccao-dbt-athena-dev-us-east-1/data/`
  * dependency dir: `s3://ccao-dbt-athena-dev-us-east-1/data/packages/<username>/`
* `ci`
  * data dir: `s3://ccao-dbt-athena-ci-us-east-1/data/`
  * dependency dir: `s3://ccao-dbt-athena-ci-us-east-1/data/packages/<branch_slug>/`
* `prod`
  * data dir: `s3://ccao-athena-ctas-us-east-1/`
  * dependency dir: `s3://ccao-athena-ctas-us-east-1/packages/`

I think `prod` is probably the most reasonable of the three. They all feel acceptable right now, but not perfect. What do you think?

File: .github/workflows/build_and_test_dbt.yaml [hunk 3]
```diff
@@ -57,6 +57,40 @@ jobs:
...
+              model_names=$(\
+                dbt list \
+                  --quiet -t "$TARGET" -s state:modified state:new \
+                  --state "$STATE_DIR" --resource-types model \
+                  --output json --output-keys name | \
+                jq -r '.name' | \
+                tr '\n' ' '
+              )
+              if [ -z "$model_names" ]; then
+                echo "No modified/new resources found, skipping dependency deploy"
```
Comment (jeancochrane) [inline .github/workflows/build_and_test_dbt.yaml:77]: See [this workflow run](https://github.com/ccao-data/data-architecture/actions/runs/9024036185/job/24797106700?pr=435#step:7:53) for an example of this conditional branch in action.

File: dbt/macros/tests/test_generate_schema_name.sql [hunk 4]
```diff
@@ -5,13 +5,6 @@
     {% do test_generate_schema_name_raises_for_default_schema_name() %}
 {% endmacro %}
 
-{% macro mock_env_var(var_name) %}
-    {% if var_name == "USER" %} {{ return("test-user") }}
-    {% elif var_name == "HEAD_REF" %} {{ return("testuser/feature-branch-1") }}
-    {% else %} {{ return("") }}
-    {% endif %}
-{% endmacro %}
```
Comment (jeancochrane) [inline dbt/macros/tests/test_generate_schema_name.sql:13]: This mock hasn't been deleted, I just moved it to the `utils` module below since I'm now reusing it outside the context of the `test_generate_schema_name` module.

File: dbt/macros/get_s3_dependency_dir.sql (hunk 2 above)
Comment (dfsnow) [reply dbt/macros/get_s3_dependency_dir.sql:40]: **suggestion (non-blocking):** Well, if we're following our existing schema for the dev, ci, prod split, something like this might be better:

* `dev`
  * data dir:  `s3://ccao-dbt-athena-dev-us-east-1/data/`
  * dependency dir: `s3://ccao-dbt-athena-dev-us-east-1/packages/<username>/`
* `ci`
  * data dir: `s3://ccao-dbt-athena-ci-us-east-1/data/`
  * dependency dir: `s3://ccao-dbt-athena-ci-us-east-1/packages/<branch_slug>/`
* `prod`
  * data dir: `s3://ccao-athena-ctas-us-east-1/`
  * dependency dir: `s3://ccao-athena-packages-us-east-1/`

File: .github/scripts/deploy_dbt_model_dependencies.sh
Comment (dfsnow) [inline .github/scripts/deploy_dbt_model_dependencies.sh:1]: **praise:** This is some pretty wild bash. Appreciate the clear comments!

**Review by dfsnow — APPROVED** [review]
This is awesome @jeancochrane. Just a minor issue and a suggestion and I think this is set to go.

File: dbt/macros/get_s3_dependency_dir.sql (hunk 2 above)
Comment (jeancochrane) [reply dbt/macros/get_s3_dependency_dir.sql:40]: @dfsnow I agree, that pattern would be cleaner, but [per my comment above](https://github.com/ccao-data/data-architecture/pull/435/files#r1596033934) we aren't actually able to configure the dependency directory as a dedicated profile variable, so it needs to be derivable from other variables like `s3_data_dir`. This is why I proposed the organization sketched out in this PR, as imperfect as it is; if you have suggestions for a better one that is derivable from existing profile variables I would be grateful!

[the remark this answers is not in the document]
Comment (jeancochrane) [reply dbt/macros/get_s3_dependency_dir.sql:40]: Good call, I fixed handling for all-space variables and added tests in 9dde41d5cd69f3067085ee2457e64fd8206cf9e8!

Comment (jeancochrane) [reply dbt/macros/get_s3_dependency_dir.sql]: I like that idea! Done in 02cc83a3ee89b3bda571f489d9f04a125ab069b3.

**Comment by jeancochrane** [comment]
@dfsnow Want to give this one more look before I merge?

File: .github/scripts/cleanup_dbt_resources.sh [hunk 5]
```diff
@@ -51,5 +51,12 @@ echo "$schemas"
 
 echo "$schemas" | xargs -i bash -c 'delete_database {}'
 
+s3_dependency_dir=$(dbt run-operation print_s3_dependency_dir --quiet \
+    --target "$1"
+)
+
+echo "Deleting Python model requirements at $s3_dependency_dir"
+aws s3 rm "$s3_dependency_dir" --recursive
```
[the remark this answers is not in the document]
Comment (dfsnow) [reply .github/scripts/cleanup_dbt_resources.sh:59]: Hmmm alright, as long as there's no way for that call to ever target anything other than the target env buckets then I think it's okay. I can dig a little more on this when we do our permissions cleanup.

File: dbt/models/default/schema/default.vw_pin_address.yml [hunk 6]
```diff
@@ -1,6 +1,11 @@
 models:
   - name: default.vw_pin_address
     description: '{{ doc("view_vw_pin_address") }}'
+    config:
+      packages:
+        - "assesspy==1.1.*"
+        - "numpy==1.26.*"
+        - "pandas==2.*"
```
Comment (dfsnow) [inline dbt/models/default/schema/default.vw_pin_address.yml:8]: **question (blocking):** Why are these here? Just for testing?

**Review by dfsnow — APPROVED** [review]

File: dbt/models/default/schema/default.vw_pin_address.yml (hunk 6 above)
Comment (jeancochrane) [reply dbt/models/default/schema/default.vw_pin_address.yml]: Sorry about that, this was one of the things that prompted me to un-request your review after realizing I had requested it prematurely! I'll just DM you next time that happens instead of messing around with the GitHub review request settings. Fixed in [a5f5262](https://github.com/ccao-data/data-architecture/pull/435/commits/a5f5262cc49cff8db3659e31b1cab979588b44dc).

File: .github/scripts/cleanup_dbt_resources.sh (hunk 5 above)
[the remark this answers is not in the document]
Comment (jeancochrane) [reply .github/scripts/cleanup_dbt_resources.sh:59]: Sounds good! I think we're also slightly more protected by the switch to variables for defining `s3_dependency_dir` in [02cc83a](https://github.com/ccao-data/data-architecture/pull/435/commits/02cc83a3ee89b3bda571f489d9f04a125ab069b3), since now our S3 paths for dependencies don't overlap with our paths for data.
